### ISSUE TITLE
Improve EIP/VIP failover when the node running CPEM fails

### ISSUE THREAD

**Issue by cprivitere** [equinix/cloud-provider-equinix-metal#304]
Right now in a cluster with an HA control plane using CPEM for EIP management you can have the following situation.

### Situation
3-controlplane-node Kubernetes cluster, created with CAPP. Has CPEM, with the control plane EIP pointing at one of the nodes (not kube-vip or metallb, but rather CPEM’s direct control), CPEM running on the same node that EIP is pointing to.
CP nodes
  A    <--- EIP ; CPEM here
  B
  C (edited) 

Node A goes away.

CPEM should move the EIP to node B or C. But CPEM is gone too, so it cannot do so until k8s controllers move it to node B.
Problem: It never happens. Controller and scheduler never seem to learn that it has something to do, CPEM stays dead, EIP never moves, we are stuck.
What does work: Each CP node’s controller-manager and scheduler do point at their local IP kube-apiserver, which in turn points at the local etcd.

###Root cause
kubelet. /etc/kubernetes/kubelet.conf points to the EIP:
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data:  // etc. etc.
    server: https://1.2.3.4:6443/      // <===== this is the EIP, not the local
  name: default-cluster
```

So all of the components are working, but the kubelets cannot connect to the apiserver, so controllers do not actually know what the state of anything is.

**Comment by cprivitere** [comment]
I did investigation into what options cluster-api lets you pass down to kubeadm configuration for changing what IPs the kubelet is configured to use. Not only is there nothing that lets you override the address kubelet talks to with cluster-api OR kubeadm, I eventually found this:

**Note**: kubeadm applies the same KubeletConfiguration to all nodes in the cluster. To apply node specific settings you can use kubelet flags as overrides by passing them in the nodeRegistration.kubeletExtraArgs (i’ve checked, there are no flags that help us) field supported by both InitConfiguration and JoinConfiguration. Some kubelet flags are deprecated, so check their status in the [kubelet reference documentation](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet) before using them.

So, the long and short is that my research is indicating that overriding the IP that kubelet is set to talk to just for the controplanes isn’t really a normal pattern (at least not for kubeadm).

I’m concerned we need to look more at changing how CPEM runs or just go with a hack that plumbs the main EIP onto each control planes localhost interface.

**Comment by cprivitere** [comment]
I took a look at a bunch of the k8s vendors to see what they all did for their cloud providers.

### Daemonset
AWS
OpenStack
Linode
Azure (has a node manager daemonset as well)
Vultr
OpenStack
vSphere
alibaba
Oracle
### Deployment
Equinix Metal
DigitalOcean
Azure (deployment with multiple replicas or static pods on each CP)
IBM (3 replicas)

**Comment by cprivitere** [comment]
@deitch has noted:

We once looked at having it in 2 modes - DaemonSet and Deployment - simultaneously; Deployment for typical CCM functions, DaemonSet for IP control. But we really didn’t want to get into the maintenance and support headaches.

So going the daemonset or static pod route could be biting off more than we can chew.

**Comment by cprivitere** [comment]
Alternately, it could improve end user reliability to move people away from CPEM managed EIPs and to use kube-vip instead.

**Comment by deitch** [comment]
Some extensive updates. @cprivitere rightly pointed out that many other CCMs run either as a `DaemonSet`, or as a `Deployment`/`StatefulSet` with `replicas` > 1. We always were hesitant about doing this due to worries about race conditions.

It turns out the `k8s.io/cloud-provider` already provides built-in leader election, for precisely this reason. This means it _should_ be safe to run multiple copies of CPEM; e.g. as a `DaemonSet` restricted to control plane nodes, as @cprivitere suggested. This is subject to testing of course. I will tackle it once we get further along with the design.

In order for this to work, there are several parts and considerations. I am putting them all together in a single list.

1. For core services (the ones supported by `k8s.io/cloud-provider`), nothing more needs to be done, other than to test it.
2. For additional services, the ones for which we run informers - notably our EIP management - we would need to get the informers from the right configuration passed to the initialization callback from `k8s.io/cloud-provider`. I do not believe we do this now, we would need to change it. We should first change it "as is" and make sure it works, and only then move to testing in parallel.
3. This all depends on leader-election, which in turn relies on access to the kubernetes service. Given that the entire purpose of this exercise is to handle failure of the control plane (kube-apiserver) node to which the EIP points, this may very well fail. On the other hand, it may very well work, since access to the `service/kubernetes` is via iptables rules. So it may be ok.

All of this leaves several layers of changes and testing:

1. Change our informers-driven "additional" components to use the informers from the provided configuration; test
2. Run multiple CPEM, test the core components for race conditions and logs to show proper behaviour
3. Run multiple CPEM, test the additional components for race conditions and logs to show proper behaviour
4. Create cluster with multiple control plane nodes and EIP pointing at one node; stop CPEM; bring down the node; test if able to access `service/kubernetes` from other nodes
5. Create cluster with multiple control plane nodes and EIP pointing at one node with CPEM on it; bring down the node; test if CPEM restarts on other nodes

I suspect that even with `service/kubernetes` working, this will fail, as it depends on kubelet connections to `kube-apiserver`, but let's see.

**Comment by deitch** [comment]
> Run multiple CPEM, test the core components for race conditions and logs to show proper behaviour
> Run multiple CPEM, test the additional components for race conditions and logs to show proper behaviour

OK, confirmed that this works. It works even without changing our informers, but we should do so anyways, as without it, it might fail when leader election switches things around.

**Comment by deitch** [comment]
> Create cluster with multiple control plane nodes and EIP pointing at one node; stop CPEM; bring down the node; test if able to access service/kubernetes from other nodes

Confirmed this does not work. It tries to access the service represented by  the Service `default/kubernetes` in the cluster, but it fails, as that routes to the primary node. It works fine if the CPEM is on a node _other_ than the node that the EIP points at (which we knew already), but if on the same node, CPEM will not bring the EIP around, even if multiple running, because it does not know that the node is gone.

It does not even work correctly - unless you catch the timing just right - if CPEM is on a different node, as it depends on the node event to determine that it is time to move the EIP over.

Basically, CCM is architected around an assumption that the `default/kubernetes` service always will be available via routing somehow, at a level beyond CPEM.

We can fix this partially by:

1. Having multiple CPEM run at once (run it as a `DaemonSet`, as @cprivitere suggested)
2. Adding a separate control loop, completely independent of the usual loops or informer events, that checks the EIP status.

We had something like that a while ago, and replaced it with this generally more robust solution, but now have a case where it doesn't completely work.

This won't completely work, though. Because the entire startup depends on leader election, if the node with both CPEM and the EIP fails, then we end up with leader election stuck, and the separate loop will not run. The above partial solution resolves the CPEM-on-different-node-than-EIP issue, not CPEM-on-same-node-as-EIP issue.

Maybe it is time to say goodbye to CPEM-managed EIP? Or to have it as a separate DaemonSet container that specializes?

**Comment by deitch** [comment]
> we should do so anyways, as without it, it might fail when leader election switches things around

Deeper investigation confirms that we actually are doing this correctly. Our pattern is a little different than the one used by Azure and AWS, but from the same basic clientbuilder. So we are fine there.

The cases to be addressed (each more complex than the previous):

1. Node dies, contains neither CPEM nor primary EIP: no problem at all.
2. Node dies, contains CPEM but **not** primary EIP: with apiserver still working, pod controller should start a new CPEM on a different node. However, it _will_ take time, as without CPEM, k8s doesn't know if it is a temporary or permanent loss, and waits. This can be addressed by switching to a `DaemonSet` that runs on all control plane nodes. This has 2 benefits:
   * leader election will cause a different CPEM instance to take over immediately
   * the new instance will immediately see the node failure is permanent and inform k8s
3. Node dies, contains CPEM **and** primary EIP: nothing we have done will help. Leader election will fail (no k8s API available), so no new CPEM, so no switching EIP, so no access to apiserver for moving pods over.

I will open a PR for `DaemonSet` to help with issue 2; for issue 3, we need some better ideas.

**Comment by displague** [comment]
Let's open a new issue for 2 and close this issue if the solution for case 3 is working (it seems this was the initial issue ask).
